### PR TITLE
Mp save refactoring

### PR DESCRIPTION
**Pull request by shmish111**
Detecting if a project has unsaved changes and dealing with it correctly is quite tricky. We were in a situation where we had a lot of state duplicated so I have managed to remove that duplication. In addition there was an issue where it is difficult to know when a user has made a change to blockly or if code was moved from the marlowe editor. In the case where code is moved from marlowe editor to blockly, we don't want to mark as having unsaved changes (because the marlowe editor and blockly are considered "views" of the same code) however if the user changes something then of course we do. Since blockly was doing loads of stuff asynchronously and since the `finished_loading` doesn't actually mean that loading has finished 😒  we need a way to detect this. I realised that `ui` events are not fired when sending code from marlowe editor to blockly so we can use the first `ui` event to tell us that the user must have started doing something. Now we have a 'switch' called `State.useEvents` (better name welcome) that we turn off just before we transfer code from marlowe editor and turn back on again at the first `ui` event.

Additionally I removed the newtype wrapper of `MainFrame.State` as it was not needed any more and I also cleaned up a few things.

Deployed to https://david.marlowe.iohkdev.io/ (there should be no functional changes)

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A plutus.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [x] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: marlowe-playground-client/src/MainFrame/Types.purs [hunk 1]
```diff
@@ -188,111 +186,105 @@ _walletSlot :: SProxy "walletSlot"
...
+    , createGistResult :: WebData Gist
+    , loadGistResult :: Either String (WebData Gist)
+    , projectName :: String
+    , showModal :: Maybe ModalView
+    -- The source of truth for the unsaved change lives inside each editor
+    -- The only reason we store a copy of the state here is because of the render function
+    -- in the Mainframe view, which uses this derived state to show an unsaved indicator.
+    -- Inside the Mainframe view we can inspect the state of the submodules (Haskell/JS/Marlowe)
+    -- as their state is part of the MainFrame state, but we cannot inspect the state of the
+    -- child components Blockly and ActusBlockly as we need a Query.
```
Comment (merivale) [inline marlowe-playground-client/src/MainFrame/Types.purs:216]: I think this comment needs updating, doesn't it? If I'm following correctly, this is now the _single_ source of truth, and is no longer copied in the editor states.

File: marlowe-playground-client/src/MainFrame/State.purs [hunk 2]
```diff
@@ -409,10 +399,7 @@ handleAction settings (HandleActusBlocklyMessage (ActusBlockly.CurrentTerms flav
         Failure e -> void $ query _actusBlocklySlot unit (ActusBlockly.SetError ("Server error! " <> (showErrorDescription (runAjaxError e).description)) unit)
         _ -> void $ query _actusBlocklySlot unit (ActusBlockly.SetError "Unknown server error!" unit)
 
--- Replicate the state of unsavedChanges from the submodule/subcomponent into the MainFrame state
-handleAction _ (HandleActusBlocklyMessage _) = do
-  hasUnsavedChanges <- queryCurrentEditorForUnsavedChanges
-  assign _hasUnsavedChanges hasUnsavedChanges
+handleAction settings (HandleActusBlocklyMessage ActusBlockly.CodeChange) = assign _hasUnsavedChanges true
```
Comment (merivale) [inline marlowe-playground-client/src/MainFrame/State.purs:402]: Could be `handleAction _ (HandleActusBlocklyMessage ActusBlockly.CodeChange) = assign _hasUnsavedChanges true`.

File: marlowe-playground-client/src/Halogen/ActusBlockly.purs [hunk 3]
```diff
@@ -39,7 +39,7 @@ type State
     , generator :: Maybe Generator
     , errorMessage :: Maybe String
     , showShiny :: Boolean
-    , hasUnsavedChanges :: Boolean
+    , useEvents :: Boolean
```
Comment (merivale) [inline marlowe-playground-client/src/Halogen/ActusBlockly.purs:42]: Erm... `uiEventHasFired`? `userHasInteracted`? Meh, they're not great names either.

**Review by merivale — APPROVED** [review]
Looks like a clear improvement to me. 👍

File: marlowe-playground-client/src/Halogen/ActusBlockly.purs (hunk 3 above)
Comment (shmish111) [reply marlowe-playground-client/src/Halogen/ActusBlockly.purs:42]: I'll leave as is until inspiration strikes with something that is _definitely_ clearer

**Pull request merged**